src/NuGet.Clients/NuGet.SolutionRestoreManager/SolutionRestoreWorker.cs

```diff
@@ -208,9 +229,14 @@ private void SolutionEvents_AfterClosing()
             // on-board request onto pending restore operation
             _pendingRequests.Value.TryAdd(request);
 
+            // await completion of the requested restore operation
+            // the caller will be unblocked immediately upon
+            // cancellation request via provided token
+            var tcs = new TaskCompletionSource<bool>();
+            using (token.Register(s => ((TaskCompletionSource<bool>)s).SetResult(false), tcs))
             using (_joinableCollection.Join())
             {
-                return await (Task<bool>)pendingRestore;
+                return await await Task.WhenAny(pendingRestore.Task, tcs.Task);
```

Comment:
[Task.WhenAny](https://msdn.microsoft.com/en-us/library/hh194914(v=vs.110).aspx) returns a task that represents the completion of one of the supplied tasks. The return task's Result is the task that completed.